libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/memcpy_async.h

```diff
@@ -0,0 +1,525 @@
+    }
+
+    if (__alignment_v < _MinInterestingAlignment) [[unlikely]] {
+        return __invoke_if_applicable<_GuaranteedAlignment, _MaxInterestingAlignment, 0, 1>::__invoke(_CUDA_VSTD::forward<_Fn>(__f));
+    }
+
+    switch (__alignment_fsb) {
+#define _ADD_CASE(val)                                                                                      \
+    case val:                                                                                               \
+        return __invoke_if_applicable<                                                                      \
+            _GuaranteedAlignment,                                                                           \
+            _MaxInterestingAlignment,                                                                       \
+            _MinInterestingAlignment,                                                                       \
+            1ull << (val - 1)>::__invoke(_CUDA_VSTD::forward<_Fn>(__f))
+
+        _ADD_CASE(12);
+        _ADD_CASE(11);
+        _ADD_CASE(10);
+        _ADD_CASE(9);
+        _ADD_CASE(8);
+        _ADD_CASE(7);
+        _ADD_CASE(6);
+        _ADD_CASE(5);
+        _ADD_CASE(4);
+        _ADD_CASE(3);
+        _ADD_CASE(2);
+        _ADD_CASE(1);
+
+#undef _ADD_CASE
+    }
+
+    _LIBCUDACXX_UNREACHABLE();
+}
+
+_LIBCUDACXX_INLINE_VISIBILITY _CUDA_VSTD::size_t __get_size(_CUDA_VSTD::size_t __size) {
+    return __size;
+}
+
+template<_CUDA_VSTD::size_t _Alignment>
+_LIBCUDACXX_INLINE_VISIBILITY _CUDA_VSTD::size_t __get_size(aligned_size_t<_Alignment> __size) {
+    return __size.value;
+}
+
+_LIBCUDACXX_INLINE_VISIBILITY
+constexpr _CUDA_VSTD::integral_constant<_CUDA_VSTD::size_t, 1> __get_alignment(_CUDA_VSTD::size_t);
+
+template<_CUDA_VSTD::size_t _Alignment>
+_LIBCUDACXX_INLINE_VISIBILITY
+constexpr _CUDA_VSTD::integral_constant<_CUDA_VSTD::size_t, _Alignment> __get_alignment(aligned_size_t<_Alignment>);
+
+namespace __arch
+{
+struct __cuda_tag {};
+template<_CUDA_VSTD::size_t _ProvidedSM>
+using __cuda = __down_convertible_constant<__cuda_tag, _ProvidedSM>;
+
+template<typename _Tp, _CUDA_VSTD::size_t _RequestedSM>
+struct __is_cuda_provides_sm : _CUDA_VSTD::false_type {
+};
+
+template<_CUDA_VSTD::size_t _ProvidedSM, _CUDA_VSTD::size_t _RequestedSM>
+struct __is_cuda_provides_sm<__cuda<_ProvidedSM>, _RequestedSM> : _CUDA_VSTD::_LIBCUDACXX_BOOL_CONSTANT((_ProvidedSM >= _RequestedSM)) {
+};
+
+struct __host {
+};
+}
+
+_LIBCUDACXX_INLINE_VISIBILITY
+async_contract_fulfillment __strided_memcpy(_CUDA_VSTD::size_t __rank, _CUDA_VSTD::size_t __group_size, char * __out_ptr, const char * __in_ptr, _CUDA_VSTD::size_t __size, _CUDA_VSTD::size_t __alignment) {
+    if (__group_size == 1) {
+        memcpy(__out_ptr, __in_ptr, __size);
+    }
+    else {
+        for (_CUDA_VSTD::size_t __offset = __rank * __alignment; __offset < __size; __offset += __group_size * __alignment) {
```

Comment:
Why is `__offset` not starting at 0 ?

Reply:
This is handling cooperative copy; the thread with rank 0 will start at 0, the thread with rank 1 will start at 1, and so on. This way we have good locality of the data that threads in a specific warp are moving, which means the hardware can coalesce it. This same idea is also done with `cp.async` later down the line.